// @builder.io/react/dist/lib/src/components/builder-content.component.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importDefault(require("react"));
var sdk_1 = require("@builder.io/sdk");
var no_wrap_1 = require("./no-wrap");
var apply_patch_with_mutation_1 = require("../functions/apply-patch-with-mutation");
var variants_provider_component_1 = require("./variants-provider.component");
var BuilderContent = /** @class */ (function (_super) {
    __extends(BuilderContent, _super);
    function BuilderContent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.ref = null;
        _this.state = {
            loading: true,
            data: null,
            updates: 1,
        };
        _this.onWindowMessage = function (event) {
            var _a;
            var message = event.data;
            if (!message) {
                return;
            }
            switch (message.type) {
                case 'builder.patchUpdates': {
                    var data = message.data;
                    if (!(data && data.data)) {
                        break;
                    }
                    var patches = data.data[(_a = _this.state.data) === null || _a === void 0 ? void 0 : _a.id];
                    if (!(patches && patches.length)) {
                        return;
                    }
                    if (location.href.includes('builder.debug=true')) {
                        eval('debugger');
                    }
                    for (var _i = 0, patches_1 = patches; _i < patches_1.length; _i++) {
                        var patch = patches_1[_i];
                        apply_patch_with_mutation_1.applyPatchWithMinimalMutationChain(_this.state.data, patch);
                    }
                    _this.setState({
                        updates: _this.state.updates + 1,
                        data: _this.state.data ? __assign({}, _this.state.data) : _this.state.data,
                    });
                    if (_this.props.contentLoaded) {
                        _this.props.contentLoaded(_this.state.data.data, _this.state.data);
                    }
                    break;
                }
            }
        };
        _this.subscriptions = new sdk_1.Subscription();
        _this.firstLoad = true;
        _this.clicked = false;
        _this.trackedImpression = false;
        _this.intersectionObserver = null;
        _this.onClick = function (reactEvent) {
            // TODO: viewport scrolling tracking for impression events
            var event = reactEvent.nativeEvent;
            var content = _this.data;
            if (!content) {
                return;
            }
            if (sdk_1.builder.autoTrack) {
                _this.builder.trackInteraction(content.id, _this.renderedVairantId, _this.clicked, event, {
                    content: content,
                });
            }
            if (!_this.clicked) {
                _this.clicked = true;
            }
        };
        return _this;
    }
    Object.defineProperty(BuilderContent.prototype, "builder", {
        get: function () {
            return this.props.builder || sdk_1.builder;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BuilderContent.prototype, "renderedVairantId", {
        get: function () {
            var _a, _b;
            var id = this.props.isStatic
                ? this.builder.getCookie("builder.tests." + ((_a = this.data) === null || _a === void 0 ? void 0 : _a.id))
                : (_b = this.data) === null || _b === void 0 ? void 0 : _b.variationId;
            if (id !== null) {
                return id;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BuilderContent.prototype, "options", {
        get: function () {
            var _a;
            var options = __assign({}, (this.props.options || {}));
            if (this.props.content && !((_a = options.initialContent) === null || _a === void 0 ? void 0 : _a.length)) {
                options.initialContent = [this.props.content];
            }
            return options;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BuilderContent.prototype, "data", {
        get: function () {
            var data = this.state.data;
            return (((this.props.inline || !sdk_1.Builder.isBrowser || this.firstLoad) &&
                this.options.initialContent &&
                this.options.initialContent[0]) ||
                data);
        },
        enumerable: true,
        configurable: true
    });
    // TODO: observe model name for changes
    BuilderContent.prototype.componentDidMount = function () {
        var _a, _b;
        // Temporary to test metrics diving in with bigquery and heatmaps
        // this.builder.autoTrack = true;
        // this.builder.env = 'development';
        if (!this.props.inline || sdk_1.Builder.isEditing) {
            this.subscribeToContent();
        }
        else if (this.props.inline && ((_b = (_a = this.options) === null || _a === void 0 ? void 0 : _a.initialContent) === null || _b === void 0 ? void 0 : _b.length)) {
            var contentData = this.options.initialContent[0];
            // TODO: intersectionobserver like in subscribetocontent - reuse the logic
            this.builder.trackImpression(contentData.id, this.renderedVairantId, undefined, {
                content: contentData,
            });
        }
        if (sdk_1.Builder.isEditing) {
            addEventListener('message', this.onWindowMessage);
        }
        /// REACT15ONLY if (this.ref) { this.ref.setAttribute('builder-model', this.props.modelName); }
    };
    BuilderContent.prototype.subscribeToContent = function () {
        var _this = this;
        if (this.props.modelName !== '_inline') {
            // TODO:... using targeting...? express.request hmmm
            this.subscriptions.add(sdk_1.builder.queueGetContent(this.props.modelName, this.options).subscribe(function (matches) {
                var match = matches && matches[0];
                _this.setState({
                    data: match,
                });
                if (match && _this.firstLoad) {
                    // TODO: autoTrack
                    if (sdk_1.builder.autoTrack) {
                        var addedObserver = false;
                        if (typeof IntersectionObserver === 'function' && _this.ref) {
                            try {
                                var observer = (_this.intersectionObserver = new IntersectionObserver(function (entries, observer) {
                                    entries.forEach(function (entry) {
                                        // In view
                                        if (entry.intersectionRatio > 0 && !_this.trackedImpression) {
                                            _this.builder.trackImpression(match.id, _this.renderedVairantId, undefined, {
                                                content: _this.data,
                                            }),
                                                { content: _this.data };
                                            _this.trackedImpression = true;
                                            if (_this.ref) {
                                                observer.unobserve(_this.ref);
                                            }
                                        }
                                    });
                                }));
                                observer.observe(_this.ref);
                                addedObserver = true;
                            }
                            catch (err) {
                                console.warn('Could not bind intersection observer');
                            }
                        }
                        if (!addedObserver) {
                            _this.trackedImpression = true;
                            _this.builder.trackImpression(match.id, _this.renderedVairantId, undefined, {
                                content: match,
                            });
                        }
                    }
                    _this.firstLoad = false;
                }
                if (_this.props.contentLoaded) {
                    _this.props.contentLoaded(match && match.data, match);
                }
            }, function (error) {
                if (_this.props.contentError) {
                    _this.props.contentError(error);
                }
            }));
        }
    };
    BuilderContent.prototype.componentWillUnmount = function () {
        if (sdk_1.Builder.isEditing) {
            removeEventListener('message', this.onWindowMessage);
        }
        this.subscriptions.unsubscribe();
        if (this.intersectionObserver && this.ref) {
            this.intersectionObserver.unobserve(this.ref);
        }
    };
    BuilderContent.prototype.render = function () {
        var _this = this;
        if (this.props.dataOnly) {
            return null;
        }
        var loading = this.state.loading;
        var useData = this.data;
        var TagName = this.props.dataOnly ? no_wrap_1.NoWrap : 'div';
        if (!this.props.isStatic) {
            return (react_1.default.createElement(TagName, __assign({}, (!this.props.dataOnly && {
                ref: function (ref) { return (_this.ref = ref); },
            }), { className: "builder-content", onClick: this.onClick, "builder-content-id": useData === null || useData === void 0 ? void 0 : useData.id, "builder-model": this.props.modelName }), this.props.children(useData === null || useData === void 0 ? void 0 : useData.data, this.props.inline ? false : loading, useData)));
        }
        return (react_1.default.createElement(variants_provider_component_1.VariantsProvider, { initialContent: useData }, function (variants, renderScript) {
            return (react_1.default.createElement(react_1.default.Fragment, null, variants.map(function (content, index) {
                // default Variation is at the end, wrap the rest with template
                // TODO: IE11 don't support templates
                var Tag = index === variants.length - 1 ? react_1.default.Fragment : 'template';
                return (react_1.default.createElement(react_1.default.Fragment, { key: String((content === null || content === void 0 ? void 0 : content.id) + index) },
                    Tag !== 'template' && (renderScript === null || renderScript === void 0 ? void 0 : renderScript()),
                    react_1.default.createElement(Tag, __assign({ key: String((content === null || content === void 0 ? void 0 : content.id) + index) }, (Tag === 'template' && {
                        'data-template-variant-id': content === null || content === void 0 ? void 0 : content.id,
                    })),
                        react_1.default.createElement(TagName, __assign({}, (index === 0 &&
                            !_this.props.dataOnly && {
                            ref: function (ref) { return (_this.ref = ref); },
                        }), { className: "builder-content", onClick: _this.onClick, "builder-content-id": content === null || content === void 0 ? void 0 : content.id, "builder-model": _this.props.modelName }), _this.props.children(content === null || content === void 0 ? void 0 : content.data, _this.props.inline ? false : loading, useData)))));
            })));
        }));
    };
    return BuilderContent;
}(react_1.default.Component));
exports.BuilderContent = BuilderContent;
